pkg/sfu/streamtracker/streamtracker_frame.go

```diff
@@ -0,0 +1,177 @@
+package streamtracker
+
+import (
+	"math"
+	"time"
+
+	"github.com/livekit/livekit-server/pkg/config"
+	"github.com/livekit/protocol/logger"
+)
+
+const (
+	checkInterval       = 500 * time.Millisecond
+	staleWindowFactor   = 5
+	frameRateResolution = float64(0.01) // 1 frame every 100 seconds
+)
+
+type StreamTrackerFrameParams struct {
+	Config    config.StreamTrackerFrameConfig
+	ClockRate uint32
+	Logger    logger.Logger
+}
+
+type StreamTrackerFrame struct {
+	params StreamTrackerFrameParams
+
+	initialized bool
+
+	tsInitialized bool
+	oldestTS      uint32
+	newestTS      uint32
+	numFrames     int
+
+	lowestFrameRate   float64
+	evalInterval      time.Duration
+	lastStatusCheckAt time.Time
+}
+
+func NewStreamTrackerFrame(params StreamTrackerFrameParams) StreamTrackerImpl {
+	s := &StreamTrackerFrame{
+		params: params,
+	}
+	s.Reset()
+	return s
+}
+
+func (s *StreamTrackerFrame) Start() {
+}
+
+func (s *StreamTrackerFrame) Stop() {
+}
+
+func (s *StreamTrackerFrame) Reset() {
+	s.initialized = false
+
+	s.tsInitialized = false
+	s.oldestTS = 0
+	s.newestTS = 0
+	s.numFrames = 0
+
+	s.lowestFrameRate = 0.0
+	s.updateEvalInterval()
+	s.lastStatusCheckAt = time.Time{}
+}
+
+func (s *StreamTrackerFrame) GetCheckInterval() time.Duration {
+	return checkInterval
+}
+
+func (s *StreamTrackerFrame) Observe(hasMarker bool, ts uint32) StreamStatusChange {
+	if !s.initialized {
+		s.initialized = true
+		if hasMarker {
+			s.tsInitialized = true
+			s.oldestTS = ts
+			s.newestTS = ts
+			s.numFrames = 1
+		}
+		return StreamStatusChangeActive
+	}
+
+	if hasMarker {
+		if !s.tsInitialized {
+			s.tsInitialized = true
+			s.oldestTS = ts
+			s.newestTS = ts
+			s.numFrames = 1
+		} else {
+			diff := ts - s.oldestTS
+			if diff > (1 << 31) {
+				s.oldestTS = ts
+			}
+			diff = ts - s.newestTS
+			if diff < (1 << 31) {
+				s.newestTS = ts
+			}
+			s.numFrames++
+		}
+	}
+	return StreamStatusChangeNone
+}
+
+func (s *StreamTrackerFrame) CheckStatus() StreamStatusChange {
+	if !s.initialized {
+		// should not be getting called when not initialized, but be safe
+		return StreamStatusChangeNone
+	}
+
+	// calculate frame rate since last check
+	frameRate := float64(0.0)
+	diff := s.newestTS - s.oldestTS
+	if diff > 0 || s.numFrames > 1 {
+		if diff > s.params.ClockRate*staleWindowFactor {
+			s.params.Logger.Infow("eval window might be stale", "numFrames", s.numFrames, "timeElapsed", float64(diff)/float64(s.params.ClockRate))
+			// STREAM-TRACKER-FRAME-TODO: might need to protect against one frame, long pause and then one or more frames, i. e. window getting stale.
+			// One possible option is to reset the fps measurement variables (tsInitialized, oldestTS, newestTS, numFrames, lowestFrameRate, evelInterval)
+			// and restart the lowest frame rate calulation process.
+		}
```

Comment:
Noting a couple of bits with this marker. I could not reproduce in my testing, but something to think about.